constructed inventory, allow adding variables directly, including ansible_group_priority

bcoca opened the pull request:
implementing #76788

##### ISSUE TYPE
- Feature Pull Request

bcoca committed:
constructed inventory add priority to keyed_groups

bcoca committed:
add docs

bcoca committed:
typeo fxi

bcoca committed:
type check handling

bcoca committed:
descri be it

bcoca committed:
handle none

bcoca committed:
note

acziryak commented:
I'm waiting on this to get merged and released to enable the inventory setup that I've architected. Anything I can do to get this pushed through?

bcoca commented:
time

acziryak commented:
I took a stab at fixing an error I ran into, and adding tests.

I have the changes here in a fork that I made: https://github.com/ansible/ansible/commit/0a6fa06a021551868bf1983f2ba99abfda1c3c56

I'm not sure what the workflow would be to get this into the PR. I guess I could open a PR against the branch in your repo, which when merged would reflect in this PR? IDK, I just don't wanna mess anything up workflow-wise for you.

I mean, I haven't ever even contributed code to Ansible before, so I don't know what a typical workflow looks like, much less whatever's going on here.

---

Re: the templating of the priority variable: I don't think that would address either mine or the original issue, since we were both utilizing it to set a static priority number. Adding that wouldn't solve any issue that's currently been brought up.

---

I'm definitely lost at what you mean by "injecting `keyed_group_name` as a variable", or in what way that might be useful. Do you mean to make it a bool, as in "This group has been created by the keyed_groups functionality"? Or something completely different? Either way, I don't think any additional variable is needed to solve the issues that lead to this PR getting created.

Thanks for all the work that you do, and I hope that whatever I can do on this can save you some time and effort.

bcoca commented:
as for workflow, you can add a 'suggested commit' to this PR itself

the templating/keyed_group_name was so you could do the following:
```
keyed_groups:
  - key: tags
    prefix: tag
    default_value: "running"
    priority: "{{ 10 if 'status' in ansible_keyed_group_name else 50 }}"
```

acziryak commented:
I wasn't able to find suggested commits on this PR, so I opened https://github.com/bcoca/ansible/pull/13 with my changes. If that's not going to work, hopefully we can figure something else out.

acziryak-everhealth committed:
Cgroup priority (#13)

* Fix setting group priority and add tests

Add all the tests to ensure the overrides for priority work as
advertised, and fix the set_priority function so that it operates on a
group rather than on a string in `__init__.py`

* Implement ansible_keyed_group_name

Add more comments
Add example
Change where we're setting the group_priority to allow it to be
templated with all available variables and `ansible_keyed_group_name`

* Reverse precedence of combining variables

* Move group priority to after the `add_host` call in case the group does not exist yet

bcoca committed:
up v

acidtracks commented:
Hello,
do you have an idea when this pr will be merged please?
I need this to merge some multi environments inventories into one.

Really helpfull thanks by advance

bcoca committed:
clog

bcoca committed:
updated clog

bcoca committed:
more comments on special var

acziryak-everhealth commented on changelogs/fragments/constructed_priority.yml:
```diff
@@ -1,2 +1,3 @@
 minor_changes:
-  - Constructed inventory (and those that inherit from it) can now set group priority on keyed_groups.
+  - The constructed inventory plugin (and those that inherit from it) can now set group priority on keyed_groups.
+  - The constructed inventnory plugin group priority is templatable and can use the `ansbile_keyed_group_name` special variable to refer to it's own keyed_group entry resolution.
```
misspelling of `inventory`: "inventnory"

bcoca committed:
topey

nitzmahone commented on lib/ansible/plugins/inventory/__init__.py:
```diff
@@ -449,6 +450,19 @@ def _add_host_to_keyed_groups(self, keys, variables, host, strict=False, fetch_h
                             result_gname = self.inventory.add_group(gname)
                             self.inventory.add_host(host, result_gname)
 
+                            # Add `ansible_keyed_group_name` to the variables that we can use to template
+                            templar_variables = combine_vars(variables, {'ansible_keyed_group_name': result_gname})
+                            self.templar.available_variables = templar_variables
+                            # Template the group_priority. There are not many variables available at this point,
+                            # But what we have should be able to be used here.
+                            group_priority = self.templar.template(keyed.get('priority', None))
+                            if group_priority is not None:
+                                result_group = self.inventory.groups.get(result_gname)
+                                try:
+                                    result_group.set_priority(int(group_priority))
+                                except ValueError as e:
+                                    raise AnsibleParserError("Invalid group priority given: %s" % to_native(e))
+
```
Maybe don't do any extra work in the common case where it's not specified, get rid of Py2-isms, leaky templar changes, walrus, move the error handling into the place that's FIXME in `group.set_priority` so we're not coercing ints and error checking in multiple places:

```suggestion
                            if (priority := keyed.get('priority')) is not None:
                                with self.templar.set_temporary_context(available_variables=variables | {'ansible_keyed_group_name': result_gname}):
                                    if (priority := self.templar.template(priority)) is not None:
                                        try:
                                            self.inventory.groups.get(result_gname).set_priority(priority)
                                        except ValueError as ve:
                                            raise AnsibleParserError(f"Invalid group priority given: {ve}")
```

bcoca committed:
Update lib/ansible/plugins/inventory/__init__.py

Co-authored-by: Matt Davis <6775756+nitzmahone@users.noreply.github.com>

nitzmahone commented on changelogs/fragments/constructed_priority.yml:
```diff
@@ -0,0 +1,3 @@
+minor_changes:
+  - The constructed inventory plugin (and those that inherit from it) can now set group priority on keyed_groups.
+  - The constructed inventory plugin group priority is templatable and can use the `ansbile_keyed_group_name` special variable to refer to it's own keyed_group entry resolution.
```
```suggestion
  - The constructed inventory plugin group priority is templatable and can use the `ansible_keyed_group_name` special variable to refer to it's own keyed_group entry resolution.
```

nitzmahone commented on lib/ansible/plugins/doc_fragments/constructed.py:
```diff
@@ -58,6 +58,12 @@ class ModuleDocFragment(object):
+      priority:
+        description:
+            - Set group's priority for fine control of variable precedence.
+            - The ``ansible_keyed_group_name`` variable that reflects the current keyed_group name, is available for templating this value.
```
```suggestion
            - The ``ansible_keyed_group_name`` variable that reflects the current keyed_group name is available for templating this value.
```

nitzmahone reviewed (changes requested):
Just a couple typos, otherwise LGTM

bcoca committed:
Update changelogs/fragments/constructed_priority.yml

Co-authored-by: Matt Davis <6775756+nitzmahone@users.noreply.github.com>

bcoca committed:
Update lib/ansible/plugins/doc_fragments/constructed.py

Co-authored-by: Matt Davis <6775756+nitzmahone@users.noreply.github.com>

jborean93 commented:
Just to play devil's advocate here I'm wondering if this can be currently achieved today with a static inventory alongside the dynamic one. You can set the `ansible_group_priority: x` for the groups you know about/need to explicitly define the priority in there. This avoids adding even more complexity to the inventory group definition which is already complex enough in favour of a solution that works like static inventories do today.

Aside from that I'm not against the feature here, just wondering if there's a better way forward.

bcoca commented:
2 caveats, 1) you need an 'extra' static inventory 2) you need to know the names a priori, while we include the possibility of setting it depending on the name, you don't need to know the whole name, just enough of a pattern.

bcoca commented:
going back to drawing board a bit on this one

acziryak commented:
I'm sorry, I don't understand why this is going back to the drawing board? Is there something that I'm able to help out with.

bcoca commented:
rethinking my approach, comments have made me think this was too narrow and that a better feature might be adding 'compose' or 'vars' at group level so you can add `ansible_group_priority` there directly AND group level vars in same plugin w/o having to add static/composite vars elsewhere

bcoca committed:
lets try this

bcoca committed:
one function to rule em all

bcoca committed:
updated clog

bcoca committed:
selfish!!

nitzmahone commented on lib/ansible/plugins/inventory/constructed.py:
```diff
@@ -29,6 +29,12 @@
             default: false
             type: boolean
             version_added: '2.11'
+        group_vars:
```
This works too, but it requires repeating the group name you've theoretically already defined in `groups`. Seems like the whole thing could be made more consistent between `keyed_groups` and `groups` if the group->criteria expression mapping under the existing `groups` key was changed to optionally accept a subdict, like:
```
...
groups:
  dude:
    membership_criteria: '{{ ... }}'
    vars:
      dudevar1: hey there
      ansible_group_priority: 2
```

bcoca replied:
I thought of that, but it would either be backwards incompatible or 'dual nature'  i.e string == condition, dict = new form ... none of the options appeal to me, so i went with the simplest one to implement

nitzmahone commented on test/integration/targets/inventory_constructed/invs/3/keyed_group_str_default_value_with_priority.yml:
```diff
@@ -3,9 +3,11 @@ keyed_groups:
     prefix: a_keyed_group
-    priority: '{{ 3 if "priority_3" in ansible_keyed_group_name else 1 }}'
+    vars:
+      ansilbe_group_priority: '{{ 3 if "priority_3" in keyed_group_2 else 1 }}'
```
```suggestion
      ansible_group_priority: '{{ 3 if "priority_3" in keyed_group_2 else 1 }}'
```
Also, `Group.set_variable` would also need to be updated to template this value inline- IIRC right now it's just calling `int()`, so when you fix this to the right varname, it's going to bomb now in `set_variable` (or at least ignore the priority set).

bcoca replied:
the templating should happen on the plugin side, specifically for this variable, group object wont have access to all vars

bcoca replied:
my other thought was .. don't make it templatable at all, still debating in my head how useful that would be

nitzmahone reviewed (changes requested):
Yeah, this definitely expands the power in the right way, but some initial UX thoughts in the review comments...

bcoca committed:
Update test/integration/targets/inventory_constructed/invs/3/keyed_group_str_default_value_with_priority.yml

Co-authored-by: Matt Davis <6775756+nitzmahone@users.noreply.github.com>

bcoca committed:
quoteit

acziryak commented:
I was coming back to this, and see that one of the tests is failing. I'm not sure if this is because the name of the group name isn't being exposed when the variables are being templated or some other issue.

If it is the issue with exposing the group name being exposed when the variables are being templated, due to the refactoring of the code to do this, is that possible anymore? And if not, I might not be understanding the code correctly, then.

acziryak commented:
Patch to fix the failing yamllint test and make the conditional `ansible_group_priority` a static int:

Per https://github.com/ansible/ansible/blob/devel/lib/ansible/inventory/group.py#L252 this gets cast to an `int` without any variable interpretation. I think this is correct as had been examined before, since this one particular value determines _which_ variables that Ansible will use to template the rest of the expressions. Before this variable is set, we don't know which group's variables take precedence to resolve this conditional. Also, I'm unsure if the previous formulation takes into account whether the particular variable is defined or not; if the `else` statement will catch the exception where the `keyed_group_2` variable does not exist. Additionally, even if this was set on the particular host, we could open the possibility that two different hosts in the same group have the same `ansible_group_priority` set, which would presumably cause undefined behavior.

I think for those reasons that `ansible_group_priority` should remain restricted to a static int.

I don't believe that any of the discussions around this had brought up use cases where a variable `ansible_group_priority` was needed:
- https://github.com/ansible/ansible/issues/66913
- https://github.com/ansible/ansible/issues/76788

Of course, I may be missing something, so let me know if I'm off-base here.

```patch
From 5baf83d7b4a6097d13c6afddf593e34fe69ed2b8 Mon Sep 17 00:00:00 2001
From: Andrew Cziryak <acziryak@updox.com>
Date: Wed, 6 Mar 2024 13:38:10 -0500
Subject: [PATCH] Fix erroring test and set static group priority

---
 lib/ansible/plugins/inventory/constructed.py                    | 2 +-
 .../invs/3/keyed_group_str_default_value_with_priority.yml      | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/ansible/plugins/inventory/constructed.py b/lib/ansible/plugins/inventory/constructed.py
index d5f502e847..b2b04acf6c 100644
--- a/lib/ansible/plugins/inventory/constructed.py
+++ b/lib/ansible/plugins/inventory/constructed.py
@@ -95,7 +95,7 @@ EXAMPLES = r'''
     group_vars:
         # group names: mapped to variables for those groups
         all:
-            ntpserver: 'ntp.{{ansible_facts['domain']}}'
+            ntpserver: "ntp.{{ansible_facts['domain']}}"
         webservers:
             open_ports: 80,8080,443
             allowed_ssh_groups: webdevs,qa,ops
diff --git a/test/integration/targets/inventory_constructed/invs/3/keyed_group_str_default_value_with_priority.yml b/test/integration/targets/inventory_constructed/invs/3/keyed_group_str_default_value_with_priority.yml
index 446b6fee1e..adb2b9ac9e 100644
--- a/test/integration/targets/inventory_constructed/invs/3/keyed_group_str_default_value_with_priority.yml
+++ b/test/integration/targets/inventory_constructed/invs/3/keyed_group_str_default_value_with_priority.yml
@@ -10,4 +10,4 @@ keyed_groups:
     default_value: "priority_3"
     prefix: a_keyed_group
     vars:
-      ansible_group_priority: '{{ 3 if "priority_3" in keyed_group_2 else 1 }}'
+      ansible_group_priority: 3
--
2.39.2 (Apple Git-143)
```

acziryak commented:
Is there anything I can do to get this moving?

bcoca commented:
im back from PTO, i should get to this soon

bcoca committed:
fixes

acziryak commented:
Hate to be a nag, but any movement on this?

bcoca commented:
i hope to soon, sorry for the delay (which is more of a nag than user having to ask about status)

bcoca committed:
bumped version

acziryak commented:
FWIW, I've delayed internally the project that required this to facilitate multiple environments which was supposed to start in Jan 2024, but have had to pick it up within the past week, and am working around the DRY-ness that this would enable in hacky other ways that duplicates config and other things. I could probably still backport changes to this type of setup, but as more and more changes to our codebase get merged, that represents exponentially more work to fix.

I think the most frustrating part of this process is that there isn't a way for me to contribute directly to this PR in order to fix little silly changes like the version here: https://github.com/ansible/ansible/pull/81603/files#diff-f2eed795bc2dc4cacab570e123a3326fa15f4f4ec96e914c344ad703ce35bec8R39, which is mostly likely going to reveal more issues in the CI which I have low confidence in getting addressed in a timely manner.

If it makes sense, I will open a new PR from a fork from this repo copied to my account to stay on top of this work. The downside I see is that the conversation in this PR will be lost. I mean, I could always put a link to this PR in the new one as a reference, but that's just more digging, and we already have the conversation spread throughout several issues and this PR around this topic, regarding rationale, opinions on implementation, and original reasons for the setting, etc. 

However, if opening up a new PR owned by myself is the way that I can best shepherd this through the PR process, I will gladly do so.
